### PR TITLE
new(falco): chart v3.0.0 for Falco 0.34 (falcoctl support and more)

### LINKED ISSUE #418: [driver-loader]: simplifying the logic behind the driver-loader
**Motivation**
The `driver-loader` is an init-container that is deployed by default when a driver (module/ebpf) is used. Users can override the default configuration and disable the `driver-loader`. There are two config variables used to disable the driver loader:
https://github.com/falcosecurity/charts/blob/06225b85abae5beb77e1a11cf0d393a8b1bb8ce8/falco/values.yaml#L188-L193
The two variables are used to configure the deployment such in this case:
https://github.com/falcosecurity/charts/blob/06225b85abae5beb77e1a11cf0d393a8b1bb8ce8/falco/templates/pod-template.tpl#L133-L146
It is not enough to just set the `driver.loader.enabled=false` in order to have the volumes mounted in the falco pod but we need to set the `driver.loader.initContainer.enabled=false`. 
I see that the `driver-loader` logic is decoupled from the `initContainer` since we can have the former `enabled` and the latter `disabled`. This allows the users to provide their own logic for the `driver-loader` but for the end user, it's confusing and error-prone. The helm chart will deploy Falco but it will not start. It's not straightforward to debug such cases.

**Feature**
I think we should simplify the `driver-loader` logic in our helm chart by setting just one configuration variable that enables or disables it. We should implement some `helpers` that mount the right volumes in the right containers based on the configuration. If we really want we can still allow people to provide their own logic for the driver loader by asking them to explicitly configure the `init containers`, `volumes`, and `volumes mounts`. 
We should aim to offer a functional and straightforward experience when deploying Falco by preventing unnecessary over-engineering of the charts.

### LINKED ISSUE #419: [falco/image]: falco errors when using "falcosecurity/falco" image
**Describe the bug**
The `falcosecurity` organizaton provides two different image flavours for Falco:
* falcosecurity/falco-driver-loader
* falcosecurity/falco
For more info: https://falco.org/docs/getting-started/download/#images.

Starting from version `2.0.0` of the charts the `falcosecurity/falco-no-driver` is the default image. As far as I know, we are still supporting the `falcosecurity/falco` image in our charts.

With the latest charts, v2.2.0, I get errors when using the `falcosecurity/falco` image.
 
In the first case, using the default values of the chart and just setting the `falcosecurity/falco` image:
```bash
helm install falco \
    --set image.repository=falcosecurity/falco \
    falcosecurity/falco
```
We have both the variables `driver.loader.enabled` and `driver.loader.initContainer.enabled` set to `true`.  The init container correctly downloads/builds the kernel module for our system but the `falco-driver-loader` script runs inside the Falco container and fails with:
```
* Looking for a Falco module locally (kernel 5.4.0-113-generic)
Detected an unsupported target system, please get in touch with the Falco community
```
That happens because we are not mounting the `/etc/` folder in the Falco container causing the script to fail. And it's ok because we are using the init container:
https://github.com/falcosecurity/charts/blob/06225b85abae5beb77e1a11cf0d393a8b1bb8ce8/falco/templates/pod-template.tpl#L133-L146

In the second case, we set to `false` both  `driver.loader.enabled` and `driver.loader.initContainer.enabled`:
```bash
helm install falco \
    --set image.repository=falcosecurity/falco \
    --set driver.loader.enabled=false \
    --set driver.loader.initContainer.enabled=false \
    falcosecurity/falco
```
The Falco pods does not start, and the error is:
```
  Warning  Failed     3s                kubelet            Error: failed to generate container "f41374f044c4a77e9a052ce42445fe1b0a962e3fbde6ae25e32b49e970734b04" spec: failed to generate spec: failed to mkdir "/sys/module/falco": mkdir /sys/module/falco: operation not permitted
```
In the latest version of the chart we add a new volume mount: `/sys/module/falco`. For more info why it is needed see here: https://github.com/falcosecurity/falco/pull/2214.
The `/sys/module/falco` folder is available only after the kernel module has been loaded.

In the third case, we set `driver.loader.enabled` to `true` but disable the `initContainer`:
```
helm install falco \
    --set image.repository=falcosecurity/falco \
    --set driver.loader.enabled=true \
    --set driver.loader.initContainer.enabled=false \
    falcosecurity/falco
```
We are removing the initContainer, but the logic of the `falco-driver-loader` script inside the Falco container should be triggered. We still get the error:
```
  Warning  Failed     4s                kubelet            Error: failed to generate container "f41374f044c4a77e9a052ce42445fe1b0a962e3fbde6ae25e32b49e970734b04" spec: failed to generate spec: failed to mkdir "/sys/module/falco": mkdir /sys/module/falco: operation not permitted
```

In order to resolve the mount issue without an init container we should mount the whole `/sys/module` folder inside the Falco container. And that is not a best practice since it is mounted in ReadWrite mode.

**How to reproduce it**
Follow the steps described above.
**Expected behaviour**
The Falco pods should be up and running after the deployment.

### PR DESCRIPTION
**Pull request by loresuso**
Signed-off-by: Aldo Lacuku <aldo@lacuku.eu>
Co-authored-by: Lorenzo Susini <susinilorenzo1@gmail.com>

**What type of PR is this?**

/kind bug

/kind cleanup

/kind design

/kind documentation

/kind feature

/kind chart-release

**Any specific area of the project related to this PR?**

/area falco-chart

**What this PR does / why we need it**:

- [x] Add support for `falcoctl` init container and sidecar container. Opening this on WIP for early feedback.
- [x] Fixes #418 
- [x] Fixes #419  
- [x] Remove default rules files from the chart 
- [x] Add modern probe support (TBD see https://github.com/falcosecurity/charts/pull/459)
- [x] Upgrade to Falco 0.34

**Which issue(s) this PR fixes**:

Fixes #418 
Fixes #419

**Special notes for your reviewer**:

**Checklist**

- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] CHANGELOG.md updated

### CODE REVIEW

File: falco/templates/_helpers.tpl
Comment (loresuso) [inline falco/templates/_helpers.tpl:244]: This is the way I found to concatenate args. Don't know if there is something smarter that can be done here!

Comment (leogr) [reply falco/templates/_helpers.tpl:244]: This :point_down: should be more straightforward! :angel: 

https://github.com/falcosecurity/charts/blob/9b27a5925f4c8a4d0fd500a2e963d572e2cce767/falco/templates/pod-template.tpl#L93-L95

File: falco/templates/_helpers.tpl
Comment (leogr) [inline falco/templates/_helpers.tpl:236]: I think it's better to create a helper here. We already do that for the other images. For example :point_down: 

https://github.com/falcosecurity/charts/blob/9b27a5925f4c8a4d0fd500a2e963d572e2cce767/falco/templates/pod-template.tpl#L55

File: falco/templates/_helpers.tpl
Comment (leogr) [inline falco/templates/_helpers.tpl:269]: Same here.

File: falco/templates/_helpers.tpl
Comment (leogr) [inline falco/templates/_helpers.tpl:277]: Same here.

File: falco/templates/pod-template.tpl
Comment (leogr) [inline falco/templates/pod-template.tpl:137]: This differs from the default Falco configuration. How do we enforce this in the Falco config?

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:15]:
```suggestion
  tag: ""
```

Please don't forget to revert this.

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:403]: This should not work as expected because the `falco_rules.yaml` will be downloaded in another folder, and as result, Falco will crash (rules will be loaded twice).

Moreover, this approach is fragile (see one of the comments above, where I wonder how we can enforce `/etc/falco/falcoctl-rules`).

That being said, I don't have a solution yet :sweat_smile:

**Review by leogr — COMMENTED** [review]
Great job @loresuso ! :star_struck: 

This PR goes almost in the correct direction, there're only a couple of things to be improved, and just an issue to be addressed.

See my comments.

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:371]:
```suggestion
        refs: [ falco-rules:0 ]
```

File: falco/values.yaml
Comment (leogr) [reply falco/values.yaml:403]: Side note: order matters here. To improve the UX, we should also pay attention to loading custom user-provided rules, if any, always _after_ those already installed.

**Comment by loresuso** [comment]
@leogr we addressed all the comments. Could you please check again? Thanks!

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:206]:
```suggestion
        tag: ""
```
Same as above

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:462]:
```suggestion
  load_plugins: []
```
Revert this pls.

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:388]: This is not completely clear to me.

How could the user `change the mount path of the emptyDir inside...` since those values are hard-coded in the chart :point_down: ?

https://github.com/falcosecurity/charts/pull/453/files#diff-701550ac61b2456be7b7355b821a218487fff642078f5d0aa22f1be38c725a75R260-R263

We may use those values to actually set the mount path inside the container OR disallow the user to change those values.
I prefer the first option (IIRC, we already to the same with some other option, maybe the gRPC socket path? :thinking:). However, I haven't thoughts if there're implications.

File: falco/templates/_helpers.tpl
Comment (leogr) [inline falco/templates/_helpers.tpl:263]: Probably, `/plugins` and `/rulesfiles` shouldn't be hardcoded.
See my comment below (the one attached to the config).

File: falco/templates/pod-template.tpl
Comment (leogr) [inline falco/templates/pod-template.tpl:140]: Unless I'm missing something...
Should those paths be mounted only when falcoctl's install or follow features are enabled? :thinking:

**Review by leogr — COMMENTED** [review]
This is impressive! :star_struck: 

I've left some comments, but overall this PR looks great.

Also, please bump the chart version and add a bit of documentation :pray: 
I propose bumping the chart to 3.0.0 since the default behavior is significatively changed.

File: falco/templates/NOTES.txt
Comment (leogr) [inline falco/templates/NOTES.txt:31]: Not sure about this. k8saudit and cloudtrail are included in the container image, and it shouldn't be a hard requirement to use falcoctl to use them.

That should be working already since when both the _install_ and _follow_ are disabled, no empty dir is mounted.

File: falco/templates/pod-template.tpl
Comment (leogr) [inline falco/templates/pod-template.tpl:104]: Please, that should be documented in the changelog.

File: falco/values-k8saudit.yaml
Comment (leogr) [inline falco/values-k8saudit.yaml:25]: We may want to enable this only for `k8saudit-rules`. Perhaps, just explaining that in the comment would be enough for now.

File: falco/values-syscall-k8saudit.yaml
Comment (leogr) [inline falco/values-syscall-k8saudit.yaml:34]:
```suggestion
        refs: [falco-rules:0, k8saudit-rules:0, k8saudit:0, json:0]
```

File: falco/values-k8saudit.yaml
Comment (leogr) [inline falco/values-k8saudit.yaml:31]:
```suggestion
        refs: [k8saudit-rules:0, k8saudit:0, json:0]
```

File: falco/values-syscall-k8saudit.yaml
Comment (leogr) [inline falco/values-syscall-k8saudit.yaml:37]: Perhaps we want to add `k8saudit-rules:0` too.

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:351]: Why do we want `--verbose` by default?

File: falco/values.yaml
Comment (leogr) [inline falco/values.yaml:366]: Why do we want `--verbose` by default?

**Review by leogr — COMMENTED** [review]
Another review iteration and this PR is getting better!

I propose to retitle this PR and use it for all changes related to Falco 0.34.

Also, it misses the following:
- to bump the chart version (it's now clear to me we should set `3.0.0`
- update the changelog
- update the readme

Anyway, great job :clap:

File: falco/templates/NOTES.txt
Comment (alacuku) [reply falco/templates/NOTES.txt:31]: The warning is printed only if we are loading plugins and one of `falcoctl.artifact` or `falcoctl.install`. So yes, it does what you are saying.

File: falco/values.yaml
Comment (alacuku) [reply falco/values.yaml:351]: I don't have a strong opinion on this.

File: falco/templates/NOTES.txt
Comment (leogr) [reply falco/templates/NOTES.txt:31]: You are right!

File: falco/values-k8saudit.yaml
Comment (leogr) [inline falco/values-k8saudit.yaml:31]:
```suggestion
falcoctl:
  artifact:
    install:
      # -- Disable the init container. We do not recommend installing (or following) plugins for security reasons since they are executable objects.
      enabled: false
    follow:
      # -- Enable the sidecar container. We do not support it yet for plugins. It is used only for rules feed such as k8saudit-rules rules.
      enabled: true
  config:
    artifact:
      follow:
        # -- List of artifacts to be followed by the falcoctl sidecar.
        # Assuming plugins k8saudit v0.5.0 and json v0.6.0 shipped in-bundle with Falco 0.34
        refs: [k8saudit-rules:0.5]
```

**Review by leogr — COMMENTED** [review]
As discussed privately with @alacuku and @loresuso, the only missing thing here is a way to disallow an unwanted plugin installation (since it's not yet fully supported in our installation scenario and security concerns may apply).

Potential solution: https://github.com/falcosecurity/falcoctl/pull/251

Otherwise SMGT.

cc @LucaGuerra (since you are the release manager for Falco 0.34), 

I think it is worth delaying a release by 1 day if we are able to get this, but let's decide together.

PS

Thank you'll, this is an amazing job :pray:

**Review by leogr — DISMISSED** [review]
:+1: 

/hold 
feel free to remove the "hold" when needed
cc @loresuso @alacuku @LucaGuerra

**Comment by loresuso** [comment]
We also fixed links in the docs. Mind to reapprove @leogr? :)

**Review by leogr — APPROVED** [review]

**Comment by leogr** [comment]
/close

Closing and reopening to trigger the CI

**Pull request merged**
